## Цель работы
Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Постановка задачи
Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Основные теоретические положения



**Бинарный метод:** 

1) Записать число в бинарном виде

2) Отбросить старший бит

3) Созать вспомогательную строку, в которую подряд записывать: SX, если бит равен единице, S - нулю.

4) Произвести вычисления: S - возведение в квадрат, а X - умножение на x


**Метод множителей:**

1) $n = pq$, где $p$ - наименьший простой множитель $n$. Таким образом, $x^n$ можно найти, вычислив $(x^p)^q$

2) Если n - простое, то можно сначала вычислить $x^{n - 1}$ и умножить на $x$
 и умножить его на x
 
3) При $n = 1$ получим $x^{n}$ без вычислений.





Аддитивная цепочка для n - последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-либо двух предыдущих.

* $l(n)$ - наименьшая длина аддитивной цепочки

* $\lambda(n) =  \lfloor\log_2(n)\rfloor $

* $\nu(n)$ - вес Хэмминга (число единиц в двоичной записи числа)

Метод множителей: $ \quad l(mn) \le l(m) + l(n) $

m-арный метод: $ \quad l(n) \le m - 2 + (k + 1)t \quad [m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j}] $



### Свойства цепочек:
* Элементы строго возрастают
* Одинаковые числа в цепочке опускаются
* Пара $ (j, k), 0 \le k \le j < i $ называется шагом $i$
* Если $\exists$ более чем 1 пара $ (j, k) $, полагаем $ max \hspace{0.2cm} j $


### Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $
### Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым

### Теорема:

Если аддитивная цепочка содержит $d$ и $f = r - d$ неудвоений, то $n \le 2^{d-1} F_{f+3}$

### Следствие:

Если аддитивная цепочка содержит $f$ удвоений и $S$ малых шагов, то

${S \le f \le} {S \over {1 - \log_2(\varphi)}}$
### Алгоритм Брауэра:

Для $n \in ℕ$ при заданном $k \in ℕ$ можно построить цепочку Брауэра с помощью формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Данная цепочка имеет длину
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимализирована для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

**Алгоритм:**

* Задаётся $k$
Вычисляются вспомогательные числа:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d \longrightarrow n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d \longrightarrow q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d,$ следовательно $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ... 2^k - 1, \\
2q_s, 4q_s, 8q_s, ... 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ... 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
... \\
... 2^k q_1, 2^k q_1 + r_1 = n.$$

### Алгоритм Яо:

* Обладает такой же вычислительной мощностью, что и алгоритм Брауэра
* $k \ge 2$, $n$ раскладывается в $2^k$-ой системе счисления.
* $$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

**Алгоритм**:

* Базовая последовательность: $1, 2, 4, ..., 2^{\lambda(n)}$
* Вычисляются значения $d(z)$ для всех $z \in \{ 1, 2, ..., 2^k - 1\}, \quad d(z) \ne 0$
* Вычисляются значения $zd(z)$ для всех $z$
* n представляетсяв виде
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$


## Алгоритм дробления вектора индексов
**Алгоритм**:

1) Исходный вектор ${1, 2, ... m}$ по которому строится исходная цепочка цепочка ${a_1 = 1, a_2, ..., a_{m+1}}$

2) Если $a_{m+1}=n$ - алгоритм завершается. Иначе вектор делится на изменяемую и неизменяемую части.

3) Находятся границы $a_{max} = a_{q} \cdot 2^{n - q}$  и  $a_{min} = a_{q+1} + m - q$, где $q$ - длина неизменяемой части. Если $n\in [ a_{min}, a_{max}]$, перебираются все возможные изменяемые части.

4) Если цепочка при данной неизменяемой части не была найдена, то изменяемая часть принимает первоначальное значение, а неизменяемая декрементируется.

5) Если были перебраны все варианты обоих частей вектора, то вектор длина вектора увеличивается на единицу и принимает значение ${1, 2, ... m + 1}$

6) Алгоритм продолжается, пока не будет найдена цепочка.

7) Замечание: имеет смысл принять граничными значениями длины вектора $l \in [\lfloor\log_2(n)\rfloor, \lfloor\log_2(n)\rfloor + \nu(n)]$

## Номер 1. Цепочки для $n \geqslant 30$ :

$ n = 42 $
* Бинарный метод:
$n = 101010_2$

$x, x^2, x^4, x^5, x^10, x^20, x^21, x^42$

Итого $k = 7$

* Метод множителей:
$ n = 42 = 2\cdot3\cdot7= 2\cdot3\cdot(1+2\cdot3)$

$x^n = (((x^2)^3\cdot x)^3)^2$

Итого $k = 7$

n = 66
* Бинарный метод:
$n = 1000010_2$

$x, x^2, x^4, x^8, x^16, x^32, x^33, x^66$

Итого $k = 7$
* Метод множителей:
$n = 66 = 2\cdot3\cdot11 = 2\cdot3\cdot(1+2\cdot5)$

$x^n =((((((x^2)^2)\cdot x)^2)\cdot x)^2)^3$

Итого $k = 7$

n=157
* Бинарный метод:
$n = 10011101_2$

$x, x^2, x^4, x^8, x^9, x^18, x^38, x^39, x^78, x^156, x^157$

Итого $k = 11$

* Метод множителей:
$n = 157 = 156+1= 2^2\cdot3\cdot13+1$

$x^n =((((((x^2)^2)^3)^2)^3)^2\cdot((x^2)^2)^3)\cdot x$

Итого $k = 10$

## Номер 2. Алгоритм Яо:

In [7]:
def YAO(a, b):
    shkaf = [] #храним число в системе счесления 2^b и частные 
    tsepochka = []
    A = a
    base = 2**b
    z = [] #храним те числа которые входят запись числа в системе 2^b
    shkaf.append([a, 0]) #добавляем число и ноль в первую строку
    while A >= base: #переводим а в систему счесления
        shkaf.append([A//base, A % base])
        z.append(A % base)
        A //= base 
    z.append(shkaf[len(shkaf) - 1][0])
    power = z #запись числа в системе 
    z = list(set(z)) #удаляем копии
    z.sort()
    print("z =", z)
    d = [] #храним d(z)
    d_z = [] #храним d(z)*z
    for i in z:
        d_i = 0;
        for j in range(0, len(power)):
            if i == power[j]:
                d_i += (2**b)**j
        d.append(d_i)
        d_z.append(d_i*i)
    print("d(z) = ", d)
    print("z*d(z) = ", d_z)
    equal = 0
    for i in d_z:
        equal += i
    print(equal)

In [8]:
a = int(input())
b = int(input())
YAO(a, b)

31415
3
z = [2, 5, 6, 7]
d(z) =  [64, 512, 8, 4097]
z*d(z) =  [128, 2560, 48, 28679]
31415


   Вывод:
   
    31415
    3
    z = [2, 5, 6, 7]
    d(z) =  [64, 512, 8, 4097]
    z*d(z) =  [128, 2560, 48, 28679]
    31415

## Номер 3. Алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа:

In [20]:
def nu(n): #счет длины Хемминга
    n = Integer(n)
    return n.popcount()

In [21]:
def create_part(m, ch_part): #эта штука создает вектор по умолчанию
    vec1 = []
    vec2 = []
    for i in range(1, m + 1):
        if i <= int(ch_part):
            vec1.append(i)
        else:
            vec2.append(i)
    return [vec1, vec2]

In [22]:
def che(vec): #а эта штука - число по вектору
    line = []
    line.append(1)
    for i in range(len(vec)):
        line.append(line[i] + line[vec[i] - 1])
    return line[len(line) - 1]

In [23]:
def dec(vec):#це декрементирует вектор 
    _len = len(vec)
    for i in range(_len):
        if vec[_len - i - 1] > 1:
            vec[_len - i - 1] -= 1
            for j in range(_len - i, _len):
                vec[j] = j + 1
            break

In [24]:
def generate_part(_len, first):#це генерирует часть вектора
    part = [0]*_len
    for i in range(_len):
        part[i] = first + i
    return part

In [25]:
def get_chain(vec):#це генерирует цепочку по вектору
    line = []
    line.append(1)
    for i in range(len(vec)):
        line.append(line[i] + line[vec[i] - 1])
    return line

In [26]:
def dect(vec):# проверка возможности декрементрирования
    for i in vec:
        if i > 1:
            return False
    return True

In [27]:
def vochcyc(n):
    l_min = math.ceil(math.log2(n)) #минимум
    l_max = int(math.log2(n) + nu(n)) #максимум
    for m in range(l_min, l_max):#перебор всех длин векторов
        q = int(m/2)
        vec = create_part(m, q)
        fix = generate_part(q, 1)
        
        while True: # перебор всех возможных фикс частей
            change = generate_part(m - q, q + 1)
            bounds = [0]*2
            a = che(fix)
            bounds[0] = a + m - q
            bounds[1] = a * 2 ** (m - q)

            
            if n < bounds[0] or n > bounds[1]:
                if dect(fix) and len(fix) > 1:
                    break
                dec(fix)
                continue
                
            while True: #перебор меняющихся частей
                if che(fix + change) == n:
                    return fix + change
                if dect(change):
                    break
                dec(change)
            if dect(fix):
                break
            dec(fix)   
    print("неа")

In [32]:
n = int(input())
vova = vochcyc(n)
print("Ваш вектор:  ", vova)
print("Ваша цепочка:", get_chain(vova))

20
Ваш вектор:   [1, 2, 3, 4, 3]
Ваша цепочка: [1, 2, 4, 8, 16, 20]


Вывод: 

    20
    Ваш вектор:   [1, 2, 3, 4, 3]
    Ваша цепочка: [1, 2, 4, 8, 16, 20]

## Номер 4. Гипотеза Шольца-Брауэра:

In [33]:
for i in range(1, 13):
    k = vochcyc(i)
    _k = vochcyc(2**i - 1)
    chain = get_chain(k)
    _chain = get_chain(_k)
    print("Теорема выполнена:", len(_chain) <= i - 1 + len(chain))

Теорема выполнена: True
Теорема выполнена: True
Теорема выполнена: True
Теорема выполнена: True


Вывод:

    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True
    Теорема выполнена: True

$n$ | $l^*(n)$ | $l^*(2^n - 1)$ | $Теорема$
--- | --- | --- | --- 
1 | 1 | 1 | Выполнена
2 | 2 | 3 | Выполнена
3 | 3 | 5 | Выполнена
4 | 3 | 6 | Выполнена
5 | 4 | 8 | Выполнена
6 | 4 | 9 | Выполнена
7 | 5 | 11 | Выполнена
8 | 4 | 11 | Выполнена
9 | 5 | 13 | Выполнена
10 | 5 | 14 | Выполнена
11 | 6 | 16 | Выполнена
12 | 5 | 16 | Выполнена

Мини-выводы: 
Полученные цепочки не являются кратчайшими;
Гипотеза Шольца-Брауэра для звездных цепочек при  n∈(1,12)  верна. 

## Выводы
Сформированы представления об аддитивных цепочках, выработаны умения составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, изучен навык использования систем компьютерной математики для реализации алгоритмов.